### PR TITLE
Scripts/Wintergrasp: Fix vehicle teleporter

### LINKED ISSUE #5831: [wintergrasp] Vehicle teleporter doesnt work
on hash 1fb07dc and earlier

The vehicle teleporter that should teleport the vehicles out of the caste doesnt work

included picture
http://img842.imageshack.us/img842/5532/wowscrnshot032112183031.jpg

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/65071-wintergrasp-vehicle-teleporter-doesnt-work?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Aokromes**
**Changes proposed**:

Fix wintergrasp vehicle teleporters

**Target branch(es)**: 335/6x

**Issues addressed**: Closes #5831 

**Tests performed**: (Does it build, tested in-game, etc)

By WoWErrorDev

### CODE REVIEW

File: src/server/scripts/Northrend/zone_wintergrasp.cpp [hunk 1]
```diff
@@ -337,30 +337,48 @@ class go_wg_vehicle_teleporter : public GameObjectScript
...
             {
-                if (_checkTimer <= diff)
+                if (!cVeh->HasAura(SPELL_VEHICLE_TELEPORT))
+                    if (Vehicle* vehicle = cVeh->GetVehicleKit())
+                        if (Unit* passenger = vehicle->GetPassenger(0))
+                            if (IsFriendly(passenger))
+                                if (Creature* teleportTrigger = passenger->SummonTrigger(go->GetPositionX()-60.0f, go->GetPositionY(), go->GetPositionZ()+1.0f, cVeh->GetOrientation(), 1000))
+                                    return teleportTrigger;
+
+                return NULL;
```
Comment (Keader) [inline src/server/scripts/Northrend/zone_wintergrasp.cpp]: return nullptr;

File: src/server/scripts/Northrend/zone_wintergrasp.cpp [hunk 2]
```diff
@@ -337,30 +337,48 @@ class go_wg_vehicle_teleporter : public GameObjectScript
...
+                    _checkTimer = 0;
                 }
-                else _checkTimer -= diff;
             }
-
           private:
               uint32 _checkTimer;
         };
 
-        GameObjectAI* GetAI(GameObject* go) const override
```
Comment (Temppus) [inline src/server/scripts/Northrend/zone_wintergrasp.cpp]: why removed  override ?

File: src/server/scripts/Northrend/zone_wintergrasp.cpp [hunk 3]
```diff
@@ -337,30 +337,48 @@ class go_wg_vehicle_teleporter : public GameObjectScript
...
+
+            bool IsFriendly(Unit* passenger)
+            {
+                return ((go->GetUInt32Value(GAMEOBJECT_FACTION) == WintergraspFaction[TEAM_HORDE] && passenger->getRaceMask() & RACEMASK_HORDE) ||
+                        (go->GetUInt32Value(GAMEOBJECT_FACTION) == WintergraspFaction[TEAM_ALLIANCE] && passenger->getRaceMask() & RACEMASK_ALLIANCE));
+            }
+
 
-            void UpdateAI(uint32 diff) override
+            Creature* IsValidVehicle(Creature* cVeh)
```
Comment (Temppus) [inline src/server/scripts/Northrend/zone_wintergrasp.cpp]: I would expect IsValidVehicle returning bool. Rename to GetValidVehicle imho.

File: src/server/scripts/Northrend/zone_wintergrasp.cpp (hunk 2 above)
Comment (Aokromes) [inline src/server/scripts/Northrend/zone_wintergrasp.cpp]: blame taking ancient fixes no one cared to make a PR for years :P

File: src/server/scripts/Northrend/zone_wintergrasp.cpp [hunk 4]
```diff
@@ -337,25 +337,40 @@ class go_wg_vehicle_teleporter : public GameObjectScript
...
 
-            void UpdateAI(uint32 diff) override
+            bool IsFriendly(Unit* passenger)
             {
-                if (_checkTimer <= diff)
+                return ((go->GetUInt32Value(GAMEOBJECT_FACTION) == WintergraspFaction[TEAM_HORDE] && passenger->getRaceMask() & RACEMASK_HORDE) ||
+                        (go->GetUInt32Value(GAMEOBJECT_FACTION) == WintergraspFaction[TEAM_ALLIANCE] && passenger->getRaceMask() & RACEMASK_ALLIANCE));
+            }
+
+
```
Comment (robinsch) [inline src/server/scripts/Northrend/zone_wintergrasp.cpp]: empty line

File: src/server/scripts/Northrend/zone_wintergrasp.cpp [hunk 5]
```diff
@@ -337,25 +337,39 @@ class go_wg_vehicle_teleporter : public GameObjectScript
...
         struct go_wg_vehicle_teleporterAI : public GameObjectAI
         {
-            go_wg_vehicle_teleporterAI(GameObject* gameObject) : GameObjectAI(gameObject), _checkTimer(1000) { }
+            go_wg_vehicle_teleporterAI(GameObject* gameObject) : GameObjectAI(gameObject), _checkTimer(0) { }
 
-            void UpdateAI(uint32 diff) override
+            bool IsFriendly(Unit* passenger)
             {
-                if (_checkTimer <= diff)
+                return ((go->GetUInt32Value(GAMEOBJECT_FACTION) == WintergraspFaction[TEAM_HORDE] && passenger->getRaceMask() & RACEMASK_HORDE) ||
```
Comment (jackpoz) [inline src/server/scripts/Northrend/zone_wintergrasp.cpp]: is RaceMask the correct way to get the faction ?

Comment (tkrokli) [inline src/server/scripts/Northrend/zone_wintergrasp.cpp]: After searching in the source, I would guess no. Perhaps better to use
```cpp
return ((go->GetUInt32Value(GAMEOBJECT_FACTION) == WintergraspFaction[TEAM_HORDE] && passenger->getFaction() & HORDE) ||
        (go->GetUInt32Value(GAMEOBJECT_FACTION) == WintergraspFaction[TEAM_ALLIANCE] && passenger->getFaction() & ALLIANCE));
```
but maybe it depends on if the check needs to return a specific value or just a comparison difference (high/low).

https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Miscellaneous/SharedDefines.h#L93-L103
vs
https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Miscellaneous/SharedDefines.h#L651-L661

**Pull request merged**